include: "rules/download_data.skm"


rule all:
    input:
        expand("featureCounts/{sample}.gene_count.txt", sample=SAMPLES)
 
rule whippet:
    expand("Whippet/Quant/{sample}.psi.gz", sample=SAMPLES)
        
        
rule download_genome:
    input:
        "download/genome.download.sh"
    output:
        "Genome/dm6.fa"
#    conda:
#        "../envs/download.yaml"
    shell:
        "bash {input}"
    
rule download_sample:
    input:
        "download/{sample}.download.sh"
    output:
        "FASTQ/{sample}.fastq.gz"
#    conda:
#        "../envs/download.yaml"
    shell:
        "bash {input}"


rule hisat2_Genome_index:
    input:
        "Genome/dm6.fa"
    output:
        "Genome/Index/dm6.1.ht2"
    threads: 5
    shell:
        "hisat2-build -p 5 {input} Genome/Index/dm6"



rule hisat2_to_Genome:
    input:
        fastq = "FASTQ/{sample}.fastq.gz",
        genome = "Genome/Index/dm6.1.ht2"
    output:
        "hisat2/{sample}.sam"
    threads: 5
    shell:
        "hisat2 -p 5 -U {input.fastq} -x  Genome/Index/dm6  > {output}"



rule samTobam:
    input:
        temp("hisat2/{sample}.sam")
    output:
        "hisat2/{sample}.sorted.bam"
    shell:
        "samtools view -b  {input}  | samtools sort - -o {output} && samtools index {output}"


rule gene_count:
    input:
        gtf = "Gene_annotation/dm6.Ensembl.genes.gtf",
        bam = "hisat2/{sample}.sorted.bam"
    output:
        "featureCounts/{sample}.gene_count.txt"
    threads: 1
    shell:
        "featureCounts -a {input.gtf} -o {output} {input.bam}"




