---
title: "RNA-seq analysis in R - Instructor Materials"
author: "Stephane Ballereau, Mark Dunning, Abbi Edwards, Oscar Rueda, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_document:
    toc: yes
minutes: 300
layout: page
subtitle: Counting
---


## Introduction and data import

* Today differential expression analysis of RNASeq data, in R
* Alignment done with HISAT, results in bam file for each sample
* Instructions on obtaining these materials in supplementary section


## Counting

* Before R, counting
* Alignment tells where each read is but need to summarise data accross genes or exons
* Use feature counts, part of subread package
* Also need annotation file (gtf) so feature counts knows where all your genes, UTRs etc. are
* can get from UCSC or ENSEMBL, instructions in supplementary section

* lets look at gtf
* navigate, ls and get them to display post it note
* unzip gtf file

```
cd ~/CourseMaterials/data/counts
  
head Mus_musculus.GRCm38.80.gtf
```

* basic fC run

```
  featureCounts \
      -t exon \
      -g gene_id \
      --primary \
      -a Mus_musculus.GRCm38.80.gtf \
      -o MCL1_DJ.featureCounts \
      MCL1_DJ.bam
```
* **`-t exon`** - the **feature** type to count reads against, in this case exons 
 - where you are counting in
* **`-g gene_id`** - the **attribute** type to summarise counts by, in this case the gene ID
 - what you want it to give you back
* **`--primary`** - only count primary alignment
 - counts only alignments labelled primary in the bam file
 - aligner randomly selects alignment when multimappers have the same mapping quality
 
* -a annotation file
* -o what we want output called
* our bam we want counting

* get nice output to terminal so you know its running ok
* post it note if all ok

**`featureCounts`** has many additional options that can be used to alter the ways in which it does the counting.

```
featureCounts --help
```

* -s strandedness, check library prep
* -C same chromosome, fusions and cancer

* two output files

```
cat MCL1.DJ.featureCounts.summary
```
* number assigned to genes
* unassigned - ambiguous - more than one feature in same location
             - multimapping - non primary multimappers
             - no features - alignment in region with no feature

```
head MCL1.DJ.featureCounts
```
* starts with cmd to generate, useful
* describe cols
* genes in same order as gtf, even where no counts, join multiple samples
* feature counts can do this for us
* save time, chr 15
* ls in small_bams


```
  featureCounts \
      -t exon \
      -g gene_id \
      --primary \
      -a small_bams/Mus_musculus.GRCm38.80.chr15.gtf  \
      -o GSE60450_Lactation.chr15.featureCounts \
      small_bams/MCL*.bam
```


```
head GSE60450_Lactation.chr15.featureCounts
```

* combined table with first cols the same and new counts cols

> ## Challenge {.challenge}
>
> 1. Redo the counting over the exons, rather than the genes. Use `featureCounts --help` to find the option you need to use. Make sure featureCounts outputs the results to a new file.
```
  featureCounts \
      -t exon \
      -g exon_id \
      --primary \
      -a small_bams/Mus_musculus.GRCm38.80.chr15.gtf  \
      -o GSE60450_Lactation.chr15.featureCounts \
      small_bams/MCL*.bam
```

* there are also options for what to do with overlapping reads on multiple exons in help

> 2. Redo the counting over genes, allowing for multimapping reads. Compare the results to our intial counts.
>

-M flag 
