---
title: "RNA-seq analysis in R - Instructor Materials"
subtitle: "Differential Expression of RNA-seq data"
author: "Stephane Ballereau, Mark Dunning, Abbi Edwards, Oscar Rueda, Ashley Sawle"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_notebook:
    toc: yes
  html_document:
    toc: yes
minutes: 300
layout: page
---

# Recap of pre-processing

```{r recap, eval = FALSE}
# Read the sample information into a data frame
sampleinfo <- read_tsv("data/SampleInfo.txt")

# Read the data into R
seqdata <- read_tsv("data/GSE60450_Lactation.featureCounts", comment = "#")

# Transform the data to matrix of counts
countdata <- as.data.frame(seqdata) %>% 
    column_to_rownames("Geneid") %>% # turn the geneid column into rownames
    rename_all(str_remove, ".bam") %>% # remove the ".bam" from the column names
    select(sampleinfo$Sample) %>% # keep sample columns using sampleinfo
    as.matrix()

# filter the data to remove genes with few counts
keep <- rowSums(countdata) > 5
countdata <- countdata[keep,]
```

## Load the data


```{r loadData}
# load the RData object we created in the previous session
load("Robjects/preprocessing.RData")
ls()
dim(countdata)
sampleinfo
```

# The model formula and design matrices

First load the packages we need.

```{r setup, message = FALSE}
library(tidyverse)
library(DESeq2)
```

## Create a DESeqDataSet object with the raw data

### Creating the design model formula

```{r modelForumla}
# Use the standard R 'formula' syntax for an additive model
design <- as.formula(~ CellType + Status)
```

* What does this look like as a model matrix?

```{r modelMatrix}
modelMatrix <- model.matrix(design, data = sampleinfo)
modelMatrix
```

* Set factor levels on Status to use virgin as the intercept

```{r setFactors}
sampleinfo$Status <- factor(sampleinfo$Status, 
                              levels = c("virgin", "pregnant", "lactate"))
modelMatrix <- model.matrix(design, data = sampleinfo)
modelMatrix
```

# Build a DESeq2DataSet

* DESeq2 builds the model matrix from the formula

```{r makeDDSObj}
# create the DESeqDataSet object
ddsObj <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = sampleinfo,
                              design = design)
```

# Data exploration

* Let's plot a PCA from `vst` transformed data. 
* Can you anticipate if the interaction term will be important?

```{r pcaPlot, fig.width=5, fig.height=5}
vstcounts <- vst(ddsObj, blind=TRUE)
plotPCA(vstcounts, intgroup=c("Status", "CellType"))
```

# Differential expression analysis with DESeq2

## The `DESeq2` work flow

The main `DESeq2` work flow is carried out in 3 steps:

```{r commonSizeFactors}
ddsObj <- estimateSizeFactors(ddsObj)

ddsObj <- estimateDispersions(ddsObj)

ddsObj <- nbinomWaldTest(ddsObj)
```

## The `DESeq` command

```{r theShortVersion}
# rebuild a clean DDS object
ddsObj <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = sampleinfo,
                              design = design)
# Run DESeq
ddsObj <- DESeq(ddsObj)
```

## Generate a results table

```{r resultsTable}
res <- results(ddsObj, alpha=0.05)
head(res)
```

* what contrast have we got??

### Independent filtering

* `NA` in padj - independent filtering

### The default contrast of `results`

* why "lactate vs virgin"?

```{r veiwModelMatrix}
modelMatrix
```

* `resultsNames` allows us to view available contrasts

```{r resultsNames}
resultsNames(ddsObj)
```

* rename `res`

```{r}
resLvV <- res
rm(res)
```

# Let's retrieve the results for pregant versus virgin - use `name`

```{r resultPvV}
resPvV <- results(ddsObj, 
                  name="Status_pregnant_vs_virgin", 
                  alpha = 0.05)
resPvV
```

* Let's get the top 100 genes by adjusted p-value

```{r topGenesPvV, message = F}
topGenesPvV <- as.data.frame(resPvV) %>%
    rownames_to_column("GeneID") %>% 
    arrange(padj) %>% 
    head(100)
topGenesPvV
```

> #### Challenge 1 {.challenge}
> Obtain results for luminal vs basal and find the top 200 genes.
> Call the new results object `resBvL`.

```{r solutionChallenge1}
resLvB <- results(ddsObj, name="CellType_luminal_vs_basal", alpha=0.05)
topGenesLvB <- as.data.frame(resLvB) %>%
    rownames_to_column("GeneID") %>% 
    arrange(padj) %>%
    head(200)
topGenesLvB
```

## Contrasts

* **pregnant** v **lactate** - column --> provide a contrast.

```{r makeContrast}
resultsNames(ddsObj)

resPvL <- results(ddsObj,
                  contrast=c("Status", "pregnant", "lactate"), 
                  alpha = 0.05)
resPvL
```

# Comparing two design models

* Maybe just CellType is sufficient to the model the data??

```{r compareModels}
designC <- as.formula(~ CellType )

# Compare the designs
ddsObjC <- DESeq(ddsObj, test="LRT", reduced=designC)
resCvCS <- results(ddsObjC)
resCvCS
```

> ### Challenge 2 {.challenge}
> When we looked at the PCA it did seem that an interaction model might be
> warranted. Let's test that.  
> 1.Fit a model with interaction.  
> 2. Use the LRT to compare the two models.  
> 3. Is the number of replicates good enough to include the interaction?    
> 4. Is the interaction needed in the model?  

```{r solutionChallenge2}
designI <- as.formula(~ CellType * Status)

# lets have a look at the model matrix to see the different
model.matrix(designI, data = sampleinfo)

# Build model
ddsObjI <- DESeqDataSetFromMatrix(countData = countdata,
                              colData = sampleinfo,
                              design = designI)
# Run DESeq
ddsObjI <- DESeq(ddsObjI)

# Compare the models
ddsObjI <- DESeq(ddsObjI, test="LRT", reduced=design)
resIvA <- results(ddsObjI)
head(resIvA)
length(which(resIvA$padj<=0.05))
```

# Testing log2 fold change versus a threshold

```{r lfcThresholdTest}
resPvL2 <- results(ddsObj,
                  contrast=c("Status", "pregnant", "lactate"), 
                  alpha = 0.05,
                  lfcThreshold=0.5, 
                  altHypothesis="greaterAbs")

sum(resPvL2$padj<0.05, na.rm = T)
sum(resPvL$padj<0.05 & abs(resPvL$log2FoldChange)>=2^0.5, na.rm = T)
```

There are four possible values for `altHypothesis`:

* greaterAbs - |β|>x - tests are two-tailed
* lessAbs - |β|<x - p values are the maximum of the upper and lower tests
* greater - β>x
* less - β<x
